### PR TITLE
Add automated PR comments with useful info for maintainers

### PR DESCRIPTION
**Pull request by angelikatyborska**
I would like to propose having a PR template with checklists. I am hoping to achieve a few things by doing that:

- Remind maintainers to double-check easily omitted details like typespecs or if mentoring notes need to be changed, some of those things cannot be verified automatically by a PR check.
- Guide new contributors into running automated check before GitHub Actions do and mark their PR as scary-red :).
- Guide new contributors into making the necessary non-obvious changes that otherwise would have to be pointed out by a maintainer.

On the other hand, I am a bit worried that a wall of text in a PR template can scare off a few people, especially people new to GitHub in general.

### CODE REVIEW

**Comment by iHiD** [comment]
I've written some assorted comments from my wider thinking on things like this. I don't have a "this is good/bad" take on this PR as a whole. If you and @neenjaw are happy with this, go for it :)

---

A few thoughts:
1. I think the general "here are things to check before you submit" is good
2. I think making people actually check them is annoying (so I'd remove the checkboxes)
3. I think some people will submit PRs without manually doing the CI checks. As long as the CI checks those things, it's not that bad. So I'd maybe reword to say "The tracks CI will undertake the following checks. You might like to ensure they pass before submitting:"

As a more general thought, I feel like these "walls of text" are better as conditional comments that get posted onto PRs (via a CI flow or webhook) rather than blocking the initial PR.

So a "Thanks so much for submitting a PR 💙. Here are some things to check..."

And then the passing/failing CI can always update those checkboxes automatically.

@SaschaMann worked on a general "Congrats on opening your first Exercism issue" [CI workflow](https://github.com/exercism/exercism/pull/5469/files) - this is the sort of thing I'm thinking, and also something I'd like to roll out globally across exercism in addition to track-specific ones.

**Comment by neenjaw** [comment]
I'll look at this later day, but while the thought is in my head -- I also really like the idea of a workflow that comments on new issues or PRs by first time contributors to give them a friendly (robot) welcome and then some tips and lifelines of ways to reach out if they run into problems

**Comment by neenjaw** [comment]
I would echo @iHiD's points here.  I don't mind a guideline to inform contributors of our standards / expectations for code, this seems a little unapproachable in its current form.  I think that CI is great for catching a lot of those things mentioned, and while  it is useful to run the checks locally I don't really mind if people don't and leave it up to CI to do it for them (In those cases maybe we need to just gently nudge a person to check the CI log and correct their errors before we do too rigorous of a review).

Maybe it could look more like:

```
Hey there! Thanks for choosing to contribute to the Exercism Elixir track!

Depending on the nature of your pull request, checkout these handy guidelines:
- 📖&nbsp;Adding a concept exercise
- 🏃🏻&nbsp;Adding a practice exercise
- 🔍&nbsp;Adding analyzer tests / comments

If your changes are complicated, give us some time to respond to them, but please feel free to reach out to us.
```

**Comment by iHiD** [comment]
Yeah. I like that.

And also I think it's ok to say both:
- "To make maintainers lives easier to please consider doing these things" (with a link to a section of the contributing.md)
- "If the CI checks fail, please check the following: ..." (link to a section in the same contributing.md) - I agree with Angelika's vibe about not making CI failures feel like a failure, so pre-empting with a "how to get out of jail" thing is a friendly way of doing that. But again, we could also monitor the CI failing, then post a message saying "Oh no. The CI failed. Please check the following list:" (inline list from the OP).

Students are prompted to read the contributing.md file automatically when submitting an issue or PR, so structuring the info in there is a good place for this.

**Comment by angelikatyborska** [comment]
Ok, after reading your comments, I am realizing that I actually don't care that much about providing initial feedback to new contributors. I am perfectly fine with doing that manually every time. We get so little contributions that it's totally doable. I don't like the idea of automated comments because, unlike a PR template, they are totally outside of the contributor's control and are seen after the fact (after opening the PR), and they might be annoyingly inaccurate.

My primary concern is making sure maintainers review the PRs properly. I know myself and I know that I will forget about half of the things I should check sooner or later... What would be the best way to provide a PR checklist to maintainers without having to personally approach every single one now and in the future?

Some of the checks might appear in configlet sooner or later (e.g. checking that an exercise's files in the `config.json` really exist in the exercise), some won't ever be automated because it would be too much work (checking if the mentoring notes and the analyzer need an update), some are impossible to automate because they're subjective (checking if a change to an exercise needs a change to its requirements).

**Comment by iHiD** [comment]
> What would be the best way to provide a PR checklist to maintainers without having to personally approach every single one now and in the future?

How about we auto-post a comment saying "Maintainers - please remember to check these..."

**Comment by angelikatyborska** [comment]
Took me a while to ponder about this, but let's say I'm in on the idea of automated PR comments directed to maintainers. I would like this comment to vary depending on which files get changed in the PR. How could I achieve that? Would it be a good idea to use those two GH actions together?

- https://github.com/actions/labeler to conditionally add some special PR labels
- https://github.com/marketplace/actions/label-commenter to trigger comments on the PRs when the labels are being added?

Sascha also suggested to use https://github.com/actions/github-script directly to write a custom solution.

**Comment by iHiD** [comment]
@angelikatyborska In the exercism/exercism repo we use this, which is a two-part check and might achieve that sort of thing you want (albeit with a different check - files changed rather than a user's status): https://github.com/exercism/exercism/blob/main/.github/workflows/first-time-issue.yml

(Presuming you've not done much GHA before...) Writing workflows is actually quite straight-forward. @SaschaMann knows this stuff inside-out so could help advise you on piecing something together if you've not got experience in it yet.

File: .github/pr-commenter.yml
Comment (angelikatyborska) [inline .github/pr-commenter.yml:2]: If somebody pushes more commits that change more files, `on-update: edit` will update the existing comment made by this action. There is also `on-update: recreate` which would delete the old comment and create a new one, thus triggering a new notification. Not sure which one would be better for us.

**Comment by angelikatyborska** [comment]
I wrote my own GHA that can react to file changes and post PR comments, with snippets that get chosen depending on which files are changes in the PR.

I will document and officially release my GHA before merging this PR, but I wanted to first present it here.

The **goal** 🎯  of this GHA in this repository is to remind **maintainers** about **manual** checks and changes that might be necessary when reviewing a PR. Those differ depending on what gets changed - concepts, exercises, exercise tests etc. This is about things that **cannot be automated only**. I didn't mention anything that has automated checks in place (formatting, dialyzer etc.) or will have them when configlet has all the features.

Below I would post how my proposed comment would look like if it included every single snippet :) (very unlikely)

**Comment by angelikatyborska** [comment]
Thank you for contributing to `exercism/elixir` :purple_heart:. This is an automated PR comment for the **maintainers** of this repository and you can just ignore it.

Based on the files changed in this PR, the following manual steps should be taken when reviewing the PR:

- **General steps**
    - Should this PR receive a label with a reputation modifier (`reputation/contributed_code/{minor,major}`)?

- **Any exercise changed**
    - Should the author of the PR be added to `contributors` in `<exercise>/.meta/config.json`?
    - Do the [analyzer](https://github.com/exercism/elixir-analyzer/tree/main/test/elixir_analyzer/test_suite) and the [analyzer comments](https://github.com/exercism/website-copy/tree/main/analyzer-comments/elixir) exist for this exercise? Do they need to be changed?
    - Should the design file (`<exercise>/.meta/design.md`) be updated to document new implementation decisions?

- **Practice exercise changed**
    - Should `prerequisites`, `practices`, and `difficulty` in `config.json` be updated?
    - Are the changes in accordance with [problem specifiations](https://github.com/exercism/problem-specifications/tree/main/exercises)?

- **Practice exercise tests changed**
    - Are all tests but the first one _skipped_?
    - Does `<exercise>/.meta/tests.toml` need updating?

- **Concept exercise changed**
    - Should `prerequisites` and `practices` in `config.json` be updated?
    - Does the concept introduction provide all necessary information to solve this exercise?

- **Concept exercise tests changed**
    - Are all tests _un-skipped_?

- **Concept changed**
    - Should the author of the PR be added to `contributors` in `<concept>/.meta/config.json`?
    - Does the concept introduction provide all necessary information to solve this exercise?

---
Automated comment created by [PR Commenter](https://github.com/angelikatyborska/pr-commenter-action).

**Comment by angelikatyborska** [comment]
@iHiD I would like your opinion about this idea and the wording in those comment snippets, and @neenjaw  I would like your opinion about everything :D

File: .github/pr-commenter.yml
Comment (neenjaw) [reply .github/pr-commenter.yml:2]: Do people get notified by an edit? If yes, then edit is probably fine. If no, recreate is probably better? 🤷🏻

File: .github/workflows/pr-comment.yml
Comment (neenjaw) [inline .github/workflows/pr-comment.yml:9]: OOC why does your action run on node 12 vs node 14?

**Review by neenjaw — APPROVED** [review]
Looks good to me, I like the idea. 

I think the language seems very formal and maybe too strong if it were me typing it, but I don't think it is incorrect.

Example: 
> Should the author of the PR be added to `contributors` in `<exercise>/.meta/config.json`?

I would probably write "Does the PR author need to be credited in `<exercise>/.meta/config.json` as an author or contributor?"

I also think maybe it needs some emojis somewhere to lighten the tone.  🤖  when it indicates that it is automated.  🎉  when we are thrilled that they have taken the time to contribute, etc.

File: .github/workflows/pr-comment.yml
Comment (angelikatyborska) [reply .github/workflows/pr-comment.yml:9]: I was following https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action when creating the action and copying some code from https://github.com/actions/labeler/blob/main/action.yml. They both mention node12, so assumed that's what I should do... I'm not really sure why or what would be the consequences of bumping to node14 instead. Do you know? Does it matter?

File: .github/pr-commenter.yml
Comment (angelikatyborska) [reply .github/pr-commenter.yml:2]: No, edits don't trigger notifications. I'll change it to `recreate` then 👍

File: .github/workflows/pr-comment.yml
Comment (neenjaw) [reply .github/workflows/pr-comment.yml:9]: It might not matter for this, and depending how `ncc` handles the code it might not matter at all.  I think there are a vew advantages to v14 over v12.  v14 more security updates/bug fixes.  Also features V8 8.1 javascript runtime engine, so gives native access to things like nullish coalescing `??` and optional chaining `?.`, which makes working with objects that much nicer.

**Comment by angelikatyborska** [comment]
> I think the language seems very formal

Damn, that's not what I was trying to achieve 🙈 I seem to have a problem with sounding friendly.

> I would probably write "Does the PR author need to be credited in <exercise>/.meta/config.json as an author or contributor?"

I had no idea that `should...?` sounds more formal than `does ... need to...?`. I removed all of the "should"s and added more emoji. Feel free to commit wording changes to make it sound better if there are more improvements necessary!

File: .github/workflows/pr-comment.yml
Comment (angelikatyborska) [reply .github/workflows/pr-comment.yml:9]: If this won't affect the (future) users of the action, I'll switch to the newer node.

**Comment by SaschaMann** [comment]
Perhaps you could add links to the docs that explain how much rep should be rewarded and when to add people as contributors? just a `[(docs)](...)` or `[📖](...)` behind the text.

If I was a first time contributor (in GitHub terms), I might be confused why you need to decide if I should be added as a contributor because I wouldn't know the Exercism definition for it.

(I got pinged due to a mention earlier in the thread, sorry for the unsolicited review :P)

File: .github/pr-commenter.yml
Comment (neenjaw) [inline .github/pr-commenter.yml:19]:
```suggestion
            - :trophy: Does this PR need to receive a label with a reputation modifier (`reputation/contributed_code/{minor,major}`)? (A medium reputation amount is awarded by default)
```

File: .github/pr-commenter.yml
Comment (neenjaw) [inline .github/pr-commenter.yml:36]:
```suggestion
            - :teacher: Are the changes in accordance with the community [problem specifiations](https://github.com/exercism/problem-specifications/tree/main/exercises)?
```

File: .github/workflows/pr-comment.yml
Comment (neenjaw) [reply .github/workflows/pr-comment.yml:9]: 🤷🏻 I'm not sure you have to change it at all if you're not using those features or if they are transpiled with `ncc`.  It was just a question to learn about how your action works. 😄 (It looks awesome, btw 🎉 )

File: .github/pr-commenter.yml
Comment (iHiD) [inline .github/pr-commenter.yml:26]: I'd be a fan of putting the actual emojis in rather than the GH placeholders. I find it harder to reason about `bust_in_silhouette` than 👤 (I had to type it in the comment to work out what it was). But there may be a reason that's a bad idea that I've not thought of?

File: .github/pr-commenter.yml
Comment (iHiD) [inline .github/pr-commenter.yml:19]:
```suggestion
            - :trophy: Does this PR need to receive a label with a reputation modifier (`reputation/contributed_code/{minor,major}`)? (A medium reputation amount is awarded by default, see [docs](https://github.com/exercism/docs/blob/main/product/reputation.md#4-creating-pull-requests))
```

File: .github/pr-commenter.yml
Comment (iHiD) [inline .github/pr-commenter.yml:27]:
```suggestion
            - :microscope: Does the [analyzer](https://github.com/exercism/elixir-analyzer/tree/main/test/elixir_analyzer/test_suite) output [comments](https://github.com/exercism/website-copy/tree/main/analyzer-comments/elixir) for this exercise? Do they need to be updated?
```

File: .github/pr-commenter.yml
Comment (iHiD) [inline .github/pr-commenter.yml:36]:
```suggestion
            - :teacher: Are the changes in accordance with the community-wide [problem specifiations](https://github.com/exercism/problem-specifications/tree/main/exercises)?
```

As a maintainer, how do I know this? I feel like we need something to explain this more. It might be that a good doc is not yet for this, in which case we should add a placeholder doc and link to it.

File: .github/pr-commenter.yml
Comment (iHiD) [reply .github/pr-commenter.yml:19]: I'd like to link to the docs bit of the website, rather than the repo, as:
1. That's where the docs publicly live, and are long-lasting (they 301 automatically on renames for example)
2. The repo isn't "stable" in terms of it's structure

We could leave this until launch but then update afterwards?

File: .github/pr-commenter.yml
Comment (iHiD) [inline .github/pr-commenter.yml:43]:
```suggestion
            - :white_circle: Are all tests except the first one _skipped_?
```

File: .github/pr-commenter.yml
Comment (iHiD) [inline .github/pr-commenter.yml:67]: I'm presuming based on this that you use the concept introductions as the exercise introductions? When we generalise this to other tracks, we'll need to remember that some tracks don't do this, and that this should instead be in the concept-exercise block.

File: .github/pr-commenter.yml
Comment (iHiD) [inline .github/pr-commenter.yml:1]: I think I've seen stuff like this go into `.github/config/...` before, although I may be misremembering!

**Review by iHiD — COMMENTED** [review]
This is really great. I want to get it everywhere. I've added a spluttering of comments - all minor things. Thank you for doing this 🙂

File: .github/pr-commenter.yml
Comment (angelikatyborska) [reply .github/pr-commenter.yml:19]: Yeah, sure. The snippet bodies can be updated later without any side effect if their ids stay the same.

File: .github/pr-commenter.yml
Comment (angelikatyborska) [reply .github/pr-commenter.yml:67]: Yes, we use word-for-word exactly the same concept and exercise introduction. We'll make it explicit by using the template once configlet supports it.

File: .github/pr-commenter.yml
Comment (angelikatyborska) [reply .github/pr-commenter.yml:27]: Your proposed change makes it should like only the comments need updating, but I want to make it clear that the analyzer itself might need changing too.

File: .github/pr-commenter.yml
Comment (angelikatyborska) [reply .github/pr-commenter.yml]: Done! I am not aware of any reasons against doing that.

File: .github/pr-commenter.yml
Comment (angelikatyborska) [reply .github/pr-commenter.yml:1]: I followed the example of the labeler action when choosing the location for the config: https://github.com/actions/labeler/blob/main/action.yml#L9

Comment (iHiD) [reply .github/pr-commenter.yml:1]: OK cool. I probably misremembered then :)

**Pull request merged**
